# 2.4 Lectura y escritura de ficheros

Notebook de prácticas para alumnado de 3º de Biotecnología.

## Objetivos

- Comprender el proceso de apertura, lectura y cierre de ficheros.
- Abrir ficheros con `open()` y leerlos línea a línea con bucles.
- Usar bloques `with` y entender por que son una buena práctica.
- Limpiar y transformar texto con `strip()`, `split()`, `upper()`, `lower()` y métodos relacionados.
- Escribir ficheros de texto plano línea a línea.
- Aplicar estas tecnicas a ficheros biológicos reales: `fastq`, `fasta`, `gff3`, `csv`, `txt` y `fastq.gz`.
- Combinar lectura de ficheros con listas, diccionarios, tuplas, sets, condicionales y bucles.

## Ficheros usados en esta práctica

Todos los datos reales usados aquí están en `data/raw/`:

- `pacbio_1.fastq`
- `proteinas_bacteria.faa`
- `genes_bacteria.ffn`
- `anotacion.gff3`
- `anotacion_resumen.txt`
- `taxonomy_GTDB.csv`
- `sample2_R1.fastq.gz`
- `sample2_R2.fastq.gz`

Las salidas que escribas en los ejercicios pueden guardarse en `data/processed/`.

## Configuración del directorio de trabajo

Si Jupyter se abre desde un directorio distinto al del repositorio, las rutas relativas como `data/raw/...` pueden fallar. Ejecuta esta celda una vez al principio para que la notebook se coloque automaticamente en la raíz del repositorio.

In [1]:
from pathlib import Path
import os

candidatos = [Path.cwd().resolve(), *Path.cwd().resolve().parents]
repo_dir = None

for candidato in candidatos:
    if (candidato / "data" / "raw").exists():
        repo_dir = candidato
        break

if repo_dir is None:
    raise FileNotFoundError("No se ha encontrado la raiz del repositorio con la carpeta data/raw")

os.chdir(repo_dir)
print("Directorio de trabajo configurado en:")
print(Path.cwd())

Directorio de trabajo configurado en:
/home/carlos/Documentos/Documentos/OPOSICIONES_2025/AYUDANTE_DOCTOR_UPV_MARZO2026/Notebooks_AMDB2026_MoraMartinez


## 1. Apertura y lectura basica de ficheros

In [2]:
# Ejemplo 1: abrir, leer una linea y cerrar manualmente
ruta = "data/raw/anotacion_resumen.txt"
f = open(ruta, "r")
primera_linea = f.readline()
print(primera_linea)
f.close()

Sequence(s):



In [3]:
# Ejemplo 2: leer linea a linea con un bucle
f = open("data/raw/anotacion_resumen.txt", "r")
for linea in f:
    print(linea.strip())
f.close()

Sequence(s):
Length: 5870008
Count: 15
GC: 45.2
N50: 3044961
N ratio: 0.0
coding density: 91.0

Annotation:
tRNAs: 90
tmRNAs: 1
rRNAs: 22
ncRNAs: 3
ncRNA regions: 19
CRISPR arrays: 1
CDSs: 5130
pseudogenes: 25
hypotheticals: 196
signal peptides: 0
sORFs: 2
gaps: 0
oriCs: 1
oriVs: 0
oriTs: 0

Bakta:
Software: v1.8.2
Database: v5.0, full
DOI: 10.1099/mgen.0.000685
URL: github.com/oschwengers/bakta


### Ejercicio 1
Abre `data/raw/anotacion_resumen.txt`, lee solo la primera línea con `readline()` e imprimela. Cierra después el fichero manualmente.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 2
Abre `data/raw/anotacion_resumen.txt`, recorre sus líneas con un `for`, imprime cada línea sin salto adicional y cierra el fichero manualmente.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 3
Cuenta cuantas líneas tiene `data/raw/anotacion_resumen.txt` usando un bucle.

In [ ]:
# Escribe aqui tu solucion


## 2. Bloques `with` y limpieza de texto

Usar `with` es una buena práctica porque el fichero se cierra automaticamente aunque ocurra un error.

In [4]:
# Ejemplo 3: with
with open("data/raw/anotacion_resumen.txt", "r") as f:
    for linea in f:
        print(linea.strip())

Sequence(s):
Length: 5870008
Count: 15
GC: 45.2
N50: 3044961
N ratio: 0.0
coding density: 91.0

Annotation:
tRNAs: 90
tmRNAs: 1
rRNAs: 22
ncRNAs: 3
ncRNA regions: 19
CRISPR arrays: 1
CDSs: 5130
pseudogenes: 25
hypotheticals: 196
signal peptides: 0
sORFs: 2
gaps: 0
oriCs: 1
oriVs: 0
oriTs: 0

Bakta:
Software: v1.8.2
Database: v5.0, full
DOI: 10.1099/mgen.0.000685
URL: github.com/oschwengers/bakta


In [5]:
# Ejemplo 4: strip, split, upper y lower
texto = "  GeneA\tGeneB\tGeneC\n"
limpio = texto.strip()
partes = limpio.split("\t")
print(limpio)
print(partes)
print(partes[0].upper())
print(partes[1].lower())

GeneA	GeneB	GeneC
['GeneA', 'GeneB', 'GeneC']
GENEA
geneb


### Ejercicio 4
Repite la lectura de `anotacion_resumen.txt` usando un bloque `with`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 5
Lee `anotacion_resumen.txt` con `with` y muestra solo las líneas no vacías.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 6
Toma una línea con formato `"GC: 45.2"`, separala con `split(":")`, elimina espacios con `strip()` y convierte el valor numérico a `float`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 7
Lee `taxonomy_GTDB.csv`, ignora la primera línea y convierte los nombres de genero de las siguientes líneas a minusculas antes de imprimirlos.

In [ ]:
# Escribe aqui tu solucion


## 3. Lectura de un fichero resumen y conversion a diccionario

In [6]:
# Ejemplo 5: convertir un resumen simple en diccionario
resumen = {}

with open("data/raw/anotacion_resumen.txt", "r") as f:
    for linea in f:
        linea = linea.strip()
        if not linea or linea.endswith(":"):
            continue
        clave, valor = linea.split(":", maxsplit=1)
        resumen[clave.strip()] = valor.strip()

print(resumen)

{'Length': '5870008', 'Count': '15', 'GC': '45.2', 'N50': '3044961', 'N ratio': '0.0', 'coding density': '91.0', 'tRNAs': '90', 'tmRNAs': '1', 'rRNAs': '22', 'ncRNAs': '3', 'ncRNA regions': '19', 'CRISPR arrays': '1', 'CDSs': '5130', 'pseudogenes': '25', 'hypotheticals': '196', 'signal peptides': '0', 'sORFs': '2', 'gaps': '0', 'oriCs': '1', 'oriVs': '0', 'oriTs': '0', 'Software': 'v1.8.2', 'Database': 'v5.0, full', 'DOI': '10.1099/mgen.0.000685', 'URL': 'github.com/oschwengers/bakta'}


### Ejercicio 8
Lee `anotacion_resumen.txt` y construye un diccionario con las líneas que tengan formato `clave: valor`, ignorando encabezados y líneas vacías.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 9
A partir del diccionario del ejercicio anterior, convierte a `int` o `float` los campos numéricos que puedas identificar.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 10
Usa el diccionario generado para imprimir un pequeño resumen por pantalla con longitud total, GC y N50.

In [ ]:
# Escribe aqui tu solucion


## 4. Lectura de CSV y trabajo con cabeceras

In [7]:
# Ejemplo 6: leer cabecera y separar columnas manualmente
with open("data/raw/taxonomy_GTDB.csv", "r") as f:
    cabecera = f.readline().strip().split(",")
    primera_fila = f.readline().strip().split(",")

print(cabecera)
print(primera_fila)
print(dict(zip(cabecera, primera_fila)))

['ident', 'gtdb_representative', 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']
['GCF_000485535.1', 'f', 'd__Archaea', 'p__Halobacteriota', 'c__Halobacteria', 'o__Halobacteriales', 'f__Halalkalicoccaceae', 'g__Halalkalicoccus', 's__Halalkalicoccus jeotgali']
{'ident': 'GCF_000485535.1', 'gtdb_representative': 'f', 'superkingdom': 'd__Archaea', 'phylum': 'p__Halobacteriota', 'class': 'c__Halobacteria', 'order': 'o__Halobacteriales', 'family': 'f__Halalkalicoccaceae', 'genus': 'g__Halalkalicoccus', 'species': 's__Halalkalicoccus jeotgali'}


In [8]:
# Ejemplo 7: leer varias filas como lista de diccionarios
registros = []

with open("data/raw/taxonomy_GTDB.csv", "r") as f:
    cabecera = f.readline().strip().split(",")
    for linea in f:
        campos = linea.strip().split(",")
        registros.append(dict(zip(cabecera, campos)))

print(registros[0])
print(len(registros))

{'ident': 'GCF_000485535.1', 'gtdb_representative': 'f', 'superkingdom': 'd__Archaea', 'phylum': 'p__Halobacteriota', 'class': 'c__Halobacteria', 'order': 'o__Halobacteriales', 'family': 'f__Halalkalicoccaceae', 'genus': 'g__Halalkalicoccus', 'species': 's__Halalkalicoccus jeotgali'}
999


### Ejercicio 11
Lee la cabecera de `taxonomy_GTDB.csv` y guardala en una lista.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 12
Lee las cinco primeras filas de `taxonomy_GTDB.csv` y conviertelas en una lista de diccionarios usando `zip()`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 13
Recorre `taxonomy_GTDB.csv` y cuenta cuantas entradas pertenecen al superreino `d__Archaea`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 14
Construye un set con todos los filos distintos presentes en `taxonomy_GTDB.csv` y muestra cuantos hay.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 15
Construye un diccionario que cuente cuantas entradas hay por filo en `taxonomy_GTDB.csv`.

In [ ]:
# Escribe aqui tu solucion


## 5. Lectura de FASTA de genes y proteínas

In [9]:
# Ejemplo 8: leer cabeceras de un FASTA
cabeceras = []

with open("data/raw/proteinas_bacteria.faa", "r") as f:
    for linea in f:
        if linea.startswith(">"):
            cabeceras.append(linea.strip())

print(cabeceras[:3])

['>ELMHFE_00005 hypothetical protein', '>ELMHFE_00010 DNA primase (bacterial type)', '>ELMHFE_00015 Conjugal transfer protein TraO']


In [10]:
# Ejemplo 9: contar secuencias en un FASTA
contador = 0

with open("data/raw/genes_bacteria.ffn", "r") as f:
    for linea in f:
        if linea.startswith(">"):
            contador += 1

print(contador)

5248


In [11]:
# Ejemplo 10: leer un FASTA como diccionario simple id -> secuencia
secuencias = {}
ident = None

with open("data/raw/proteinas_bacteria.faa", "r") as f:
    for linea in f:
        linea = linea.strip()
        if linea.startswith(">"):
            ident = linea[1:].split()[0]
            secuencias[ident] = ""
        else:
            secuencias[ident] += linea

primer_id = list(secuencias.keys())[0]
print(primer_id)
print(len(secuencias[primer_id]))

ELMHFE_00005
93


### Ejercicio 16
Cuenta cuantas secuencias hay en `proteinas_bacteria.faa`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 17
Lee `genes_bacteria.ffn` y calcula la longitud de las tres primeras secuencias.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 18
Construye un diccionario `id -> descripcion` a partir de las cabeceras de `proteinas_bacteria.faa`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 19
Lee `genes_bacteria.ffn` y guarda en una lista los identificadores de las secuencias cuya longitud sea mayor que 1500 nucleotidos.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 20
Lee `proteinas_bacteria.faa` y busca cuantas secuencias contienen el motivo aminoacidico `GG`.

In [ ]:
# Escribe aqui tu solucion


## 6. Lectura de GFF3 y parseo de anotaciones

In [ ]:
# Ejemplo 11: ignorar comentarios y separar columnas
with open("data/raw/anotacion.gff3", "r") as f:
    for linea in f:
        if linea.startswith("#"):
            continue
        campos = linea.strip().split("\t")
        print(campos[:5])
        break

In [ ]:
# Ejemplo 12: contar tipos de feature
tipos = {}

with open("data/raw/anotacion.gff3", "r") as f:
    for linea in f:
        if linea.startswith("#"):
            continue
        campos = linea.strip().split("\t")
        tipo = campos[2]
        tipos[tipo] = tipos.get(tipo, 0) + 1

print(tipos)

In [ ]:
# Ejemplo 13: extraer atributos de la novena columna
with open("data/raw/anotacion.gff3", "r") as f:
    for linea in f:
        if linea.startswith("#"):
            continue
        campos = linea.strip().split("\t")
        atributos = campos[8].split(";")
        dic_atributos = {}
        for atributo in atributos:
            if "=" in atributo:
                clave, valor = atributo.split("=", maxsplit=1)
                dic_atributos[clave] = valor
        print(dic_atributos)
        break

### Ejercicio 21
Cuenta cuantas líneas de anotacion reales hay en `anotacion.gff3`, ignorando las que empiezan por `#`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 22
Construye un diccionario con el número de anotaciones por tipo de feature en `anotacion.gff3`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 23
Recorre `anotacion.gff3` y guarda en una lista los `locus_tag` de las anotaciones de tipo `CDS`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 24
Busca en `anotacion.gff3` cuantas anotaciones contienen el texto `hypothetical protein` en la columna de atributos.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 25
Calcula la longitud de cada `CDS` como `fin - inicio + 1` y guarda las 10 primeras en una lista.

In [ ]:
# Escribe aqui tu solucion


## 7. Lectura de FASTQ y análisis de calidad

En un FASTQ, cada lectura ocupa 4 líneas: cabecera, secuencia, separador `+` y calidad.

In [ ]:
# Ejemplo 14: leer una lectura FASTQ completa
with open("data/raw/pacbio_1.fastq", "r") as f:
    header = f.readline().strip()
    seq = f.readline().strip()
    plus = f.readline().strip()
    qual = f.readline().strip()

print(header)
print(len(seq), len(qual))
print(plus)

In [12]:
# Ejemplo 15: contar lecturas y longitudes
num_lecturas = 0
longitudes = []

with open("data/raw/pacbio_1.fastq", "r") as f:
    while True:
        header = f.readline().strip()
        if not header:
            break
        seq = f.readline().strip()
        f.readline()
        f.readline()
        num_lecturas += 1
        longitudes.append(len(seq))

print(num_lecturas)
print(longitudes[:5])

500
[1499, 1499, 1501, 1501, 1498]


In [14]:
# Ejemplo 16: filtrar lecturas por longitud minima
ids_largas = []

with open("data/raw/pacbio_1.fastq", "r") as f:
    while True:
        header = f.readline().strip()
        if not header:
            break
        seq = f.readline().strip()
        f.readline()
        f.readline()
        if len(seq) >= 1499:
            ids_largas.append(header)

print(len(ids_largas))

275


### Ejercicio 26
Cuenta cuantas lecturas hay en `pacbio_1.fastq`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 27
Calcula la longitud de todas las lecturas de `pacbio_1.fastq` y guarda esas longitudes en una lista.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 28
Busca cuantas lecturas de `pacbio_1.fastq` contienen el motivo `GGG` en su secuencia.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 29
Guarda en una lista los identificadores de las lecturas de `pacbio_1.fastq` cuya longitud sea mayor que 1498.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 30
Calcula la longitud media de las lecturas de `pacbio_1.fastq`.

In [ ]:
# Escribe aqui tu solucion


## 8. Lectura de FASTQ comprimido con `gzip`

In [15]:
# Ejemplo 17: leer un FASTQ comprimido
import gzip

with gzip.open("data/raw/sample2_R1.fastq.gz", "rt") as f:
    for _ in range(4):
        print(f.readline().strip())

@E100033384L1C001R00300000006/1
TGGAACATGCGCGTAAAAAGACCACTCAAAAAGTGAATCTTTTTGAGAAAGTGCAGATACCCGGATATCAGGATGCTCTGCGAAAGATAAAACGATTTATGGAGGATGAAGAGAATTTGTCCAAATCATCGCAGAAGATTCTGAAATCTA
+
FGFGGGGFGFAGG.GGGGGGGGGGGFFGGGGGGFGFG@GGGGGGGGGFGGFGGGFCFDFGFGFGG4GFGGGCEDFCGGEEDGGGFFFFGFGFCGG9EGG-GFC(EG>GBGGF@@GFEGF@GFGGF8FF?GFGGFAGFG,GGEF?GDB?FG


In [ ]:
# Ejemplo 18: contar lecturas en un FASTQ gz
contador = 0

with gzip.open("data/raw/sample2_R1.fastq.gz", "rt") as f:
    while True:
        header = f.readline().strip()
        if not header:
            break
        f.readline()
        f.readline()
        f.readline()
        contador += 1

print(contador)

In [16]:
# Ejemplo 19: filtrar por calidad muy simple usando caracteres
buenas = 0
malas = 0

with gzip.open("data/raw/sample2_R1.fastq.gz", "rt") as f:
    while True:
        header = f.readline().strip()
        if not header:
            break
        seq = f.readline().strip()
        f.readline()
        qual = f.readline().strip()
        if "!" in qual or '"' in qual:
            malas += 1
        else:
            buenas += 1

print(buenas, malas)

2500 0


### Ejercicio 31
Lee `sample2_R1.fastq.gz` con `gzip.open(..., "rt")` y muestra por pantalla la primera lectura completa.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 32
Cuenta cuantas lecturas hay en `sample2_R1.fastq.gz` y en `sample2_R2.fastq.gz`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 33
Calcula la longitud media de las secuencias en `sample2_R1.fastq.gz`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 34
Busca cuantas lecturas de `sample2_R1.fastq.gz` contienen el motivo `GAT`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 35
Haz un filtrado simple de calidad sobre `sample2_R1.fastq.gz`: considera de baja calidad las lecturas cuya cadena de calidad contenga `!` o `"`. Cuenta cuantas pasan y cuantas se descartan.

In [ ]:
# Escribe aqui tu solucion


## 9. Escritura de ficheros de texto

In [ ]:
# Ejemplo 20: escribir un fichero simple
with open("data/processed/prueba_escritura.txt", "w") as f:
    f.write("Primer linea\n")
    f.write("Segunda linea\n")

In [ ]:
# Ejemplo 21: escribir varias lineas desde una lista
genes = ["BRCA1", "TP53", "EGFR"]
with open("data/processed/lista_genes.txt", "w") as f:
    for gen in genes:
        f.write(gen + "\n")

### Ejercicio 36
Escribe un fichero `data/processed/resumen_simple.txt` con tres líneas de texto creadas por ti.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 37
Lee `anotacion_resumen.txt` y escribe en `data/processed/anotacion_resumen_mayusculas.txt` las mismas líneas, pero en mayúsculas.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 38
Escribe en un fichero nuevo solo las cabeceras de `proteinas_bacteria.faa`.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 39
Escribe un fichero con los identificadores de las lecturas de `pacbio_1.fastq` cuya longitud sea mayor que 1498.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 40
Lee `genes_bacteria.ffn` y construye un diccionario `id -> longitud`. Despues calcula cuantas secuencias tienen longitud menor de 300, entre 300 y 1000, y mayor de 1000.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 41
Lee `proteinas_bacteria.faa` y construye una lista con los identificadores de las proteínas que contengan al menos un motivo `GG`, `KK` o `RR`. Cuenta cuantas hay de cada tipo de motivo.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 42
Lee `taxonomy_GTDB.csv` y construye un diccionario `filo -> set de generos`. Despues escribe en un fichero de salida un resumen con el número de géneros distintos por filo.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 43
Lee `anotacion.gff3`, extrae solo las anotaciones `CDS`, calcula su longitud y guarda en un fichero las 20 más largas con `locus_tag`, inicio, fin y longitud.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 44
Lee `sample2_R1.fastq.gz` y `sample2_R2.fastq.gz` en paralelo. Cuenta cuantas parejas de lecturas tienen la misma longitud, cuantas no, y escribe en un fichero los identificadores de las parejas cuyas dos lecturas pasen el filtro simple de calidad.

In [ ]:
# Escribe aqui tu solucion


### Ejercicio 45
Diseña un mini pipeline de lectura de ficheros que combine al menos tres de los archivos del directorio `data/raw/`. Debe leer datos reales, construir al menos una lista y un diccionario, hacer una busqueda de patrones, generar un resumen numérico y escribir un informe final en `data/processed/`.

In [ ]:
# Escribe aqui tu solucion